### PR TITLE
feat: workflow support

### PR DESCRIPTION
**Pull request by octonato**

### CODE REVIEW

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflow/Workflow.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+import java.util.function.Function;
+
+public abstract class Workflow<S> {
+
+
+  private Optional<CommandContext> commandContext = Optional.empty();
+
+  private Optional<S> currentState = Optional.empty();
+
+  private boolean handlingCommands = false;
```
Comment (aludwiko) [inline sdk/java-sdk/src/main/java/kalix/javasdk/workflow/Workflow.java:37]: do we need it?

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala [hunk 2]
```diff
@@ -153,7 +153,7 @@ class DiscoveryImpl(
               name,
               Component.ComponentSettings.Entity(
                 EntitySettings(
-                  service.entityType,
+                  service.workflowName,
```
Comment (aludwiko) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala:156]: leftover?

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala [hunk 3]
```diff
@@ -52,7 +52,7 @@ final class EventSourcedEntityService(
     override val descriptor: Descriptors.ServiceDescriptor,
     override val additionalDescriptors: Array[Descriptors.FileDescriptor],
     val messageCodec: MessageCodec,
-    override val entityType: String,
+    override val workflowName: String,
```
Comment (aludwiko) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala:55]: leftover?

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/replicatedentity/ReplicatedEntitiesImpl.scala [hunk 4]
```diff
@@ -42,7 +42,7 @@ final class ReplicatedEntityService(
     override val descriptor: Descriptors.ServiceDescriptor,
     override val additionalDescriptors: Array[Descriptors.FileDescriptor],
     val anySupport: AnySupport,
-    override val entityType: String,
+    override val workflowName: String,
```
Comment (aludwiko) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/replicatedentity/ReplicatedEntitiesImpl.scala:45]: same here

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/valueentity/ValueEntitiesImpl.scala [hunk 5]
```diff
@@ -52,7 +52,7 @@ final class ValueEntityService(
     override val descriptor: Descriptors.ServiceDescriptor,
     override val additionalDescriptors: Array[Descriptors.FileDescriptor],
     val messageCodec: MessageCodec,
-    override val entityType: String,
+    override val workflowName: String,
```
Comment (aludwiko) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/valueentity/ValueEntitiesImpl.scala:55]: same here

File: sdk/java-sdk/src/test/java/kalix/javasdk/valueentity/CartEntity.java [hunk 6]
```diff
@@ -57,8 +57,7 @@ public Effect<Empty> addItem(
         removeItemByProductId(currentState, addLineItem.getProductId());
     lineItems.add(lineItem);
     lineItems.sort(Comparator.comparing(ShoppingCartDomain.LineItem::getProductId));
-    return effects()
-        .updateState(ShoppingCartDomain.Cart.newBuilder().addAllItems(lineItems).build())
+    return effects()        .updateState(ShoppingCartDomain.Cart.newBuilder().addAllItems(lineItems).build())
```
Comment (aludwiko) [inline sdk/java-sdk/src/test/java/kalix/javasdk/valueentity/CartEntity.java:60]: revert?

File: sdk/java-sdk/src/main/java/kalix/javasdk/workflow/Workflow.java (hunk 1 above)
Comment (octonato) [reply sdk/java-sdk/src/main/java/kalix/javasdk/workflow/Workflow.java:37]: Yes and no. 

It's wrong to call it `handlingCommands` but we do need such a flag. Accessing the `currentState` should only be possible when running a step or handling a command. Calling this method in the constructor, for instance, is illegal. 

The state is only available once we have a message to deliver to the underlying component.

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala (hunk 2 above)
Comment (octonato) [reply sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala:156]: That was certainly a too enthusiastic refactoring. 😄 

It's also not correct to call this entityType. Actions and Views are also Services, but not Entities. For Views, we are using the `viewId`, for action we are simply using the descriptor name. 

This is internal API. We can refactor it without problem. Maybe `serviceId` or `serviceName`?

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/workflow/WorkflowImpl.scala [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+      .map {
+
+        case InCommand(command) if workflowId != command.entityId =>
+          throw ProtocolException(command, "Receiving Workflow is not the intended recipient of command")
+
+        case InCommand(command) if command.payload.isEmpty =>
+          throw ProtocolException(command, "No command payload for Workflow")
+
+        case InCommand(command) =>
+          if (workflowId != command.entityId)
```
[the remark this answers is not in the document]
Comment (octonato) [reply sdk/java-sdk/src/main/scala/kalix/javasdk/impl/workflow/WorkflowImpl.scala:219]: Good catch. This is copy-n-pasta from ValueEntities. I will fix there as well.

File: sdk/java-sdk/src/test/java/kalix/javasdk/workflow/TransferWorkflow.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    var deposit =
+      step(depositStepName)
+        .call((MoneyTransferApi.Deposit cmd ) -> deferredCall(cmd, Empty.class))
+        .andThen(__ -> {
+          var state = currentState().toBuilder().setLastStep("deposited").build();
+          return effects().updateState(state).end();
+        });
+
+    return workflow("transfer-workflow")
```
Comment (aludwiko) [inline sdk/java-sdk/src/test/java/kalix/javasdk/workflow/TransferWorkflow.java:70]: Is the workflow name equivalent of `entity_type`? If yes then how this should be passed to kalix-proxy? I need this to create a persistence id, e.g. "transfer-workflow|1234-1234-1234"

Comment (octonato) [reply sdk/java-sdk/src/test/java/kalix/javasdk/workflow/TransferWorkflow.java:70]: yep, that should not be part of the WorkflowDef, but be included in the gRPC file (or annotation in Spring SDK)

Comment (octonato) [reply sdk/java-sdk/src/test/java/kalix/javasdk/workflow/TransferWorkflow.java]: name is removed now

File: sdk/java-sdk/src/test/proto/workflow/transfer_workflow.proto [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+package com.example.workflow.transfer;
+
+option java_outer_classname = "MoneyTransferApi";
+
+import "kalix/annotations.proto";
+import "google/protobuf/empty.proto";
+
+message Transfer {
+  // TODO: we should better introduce (kalix.field).workflow_id
+  string workflow_id = 1 [(kalix.field).entity_key = true];
```
Comment (aludwiko) [inline sdk/java-sdk/src/test/proto/workflow/transfer_workflow.proto:28]: I think we have `workflow_key` now?

**Review by aludwiko — APPROVED** [review]
One small comment. LGTM.

**Pull request merged**